File: ebook/src/main/java/com/artifex/mu/MuPDFCore.java
package com.artifex.mu;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.PointF;
import android.graphics.RectF;

import java.util.ArrayList;

public class MuPDFCore
{
	/* load our native library */
	private static boolean gs_so_available = false;
	static {
		System.out.println("Loading dll");
		System.loadLibrary("mupdf_java");
		System.out.println("Loaded dll");
		if (gprfSupportedInternal())
		{
			try {
				System.loadLibrary("gs");
				gs_so_available = true;
			}
			catch (UnsatisfiedLinkError e) {
				gs_so_available = false;
			}
		}
	}

	/* Readable members */
	private int numPages = -1;
	private float pageWidth;
	private float pageHeight;
	private long globals;
	private byte fileBuffer[];
	private String file_format;
	private boolean isUnencryptedPDF;
	private final boolean wasOpenedFromBuffer;

	/* The native functions */
	private static native boolean gprfSupportedInternal();
	private native long openFile(String filename);
	private native long openBuffer(String magic);
	private native String fileFormatInternal();
	private native boolean isUnencryptedPDFInternal();
	private native int countPagesInternal();
	private native void gotoPageInternal(int localActionPageNum);
	private native float getPageWidth();
	private native float getPageHeight();
	private native void drawPage(Bitmap bitmap,
			int pageW, int pageH,
			int patchX, int patchY,
			int patchW, int patchH,
			long cookiePtr);
	private native void updatePageInternal(Bitmap bitmap,
			int page,
			int pageW, int pageH,
			int patchX, int patchY,
			int patchW, int patchH,
			long cookiePtr);
	private native RectF[] searchPage(String text);
	private native TextChar[][][][] text();
	private native byte[] textAsHtml();
	private native void addMarkupAnnotationInternal(PointF[] quadPoints, int type);
	private native void addInkAnnotationInternal(PointF[][] arcs);
	private native void deleteAnnotationInternal(int annot_index);
	private native int passClickEventInternal(int page, float x, float y);
	private native void setFocusedWidgetChoiceSelectedInternal(String [] selected);
	private native String [] getFocusedWidgetChoiceSelected();
	private native String [] getFocusedWidgetChoiceOptions();
	private native int getFocusedWidgetSignatureState();
	private native String checkFocusedSignatureInternal();
	private native boolean signFocusedSignatureInternal(String keyFile, String password);
	private native int setFocusedWidgetTextInternal(String text);
	private native RectF[] getWidgetAreasInternal(int page);
	private native Annotation[] getAnnotationsInternal(int page);
	//private native OutlineItem [] getOutlineInternal();
	private native boolean hasOutlineInternal();
	private native boolean needsPasswordInternal();
	private native boolean authenticatePasswordInternal(String password);
	private native void startAlertsInternal();
	private native void stopAlertsInternal();
	private native void destroying();
	private native boolean hasChangesInternal();
	private native void saveInternal();
	private native long createCookie();
	private native void destroyCookie(long cookie);
	private native void abortCookie(long cookie);

	private native String startProofInternal(int resolution);
	private native void endProofInternal(String filename);
	private native int getNumSepsOnPageInternal(int page);
	private native int controlSepOnPageInternal(int page, int sep, boolean disable);
	private native Separation getSepInternal(int page, int sep);

	public native boolean javascriptSupported();

	public class Cookie
	{
		private final long cookiePtr;

		public Cookie()
		{
			cookiePtr = createCookie();
			if (cookiePtr == 0)
				throw new OutOfMemoryError();
		}

		public void abort()
		{
			abortCookie(cookiePtr);
		}

		public void destroy()
		{
			// We could do this in finalize, but there's no guarantee that
			// a finalize will occur before the muPDF context occurs.
			destroyCookie(cookiePtr);
		}
	}

	public MuPDFCore(Context context, String filename) throws Exception
	{
		globals = openFile(filename);
		if (globals == 0)
		{
			throw new Exception(String.format("Não é possível abrir o item selecionado", filename));
		}
		file_format = fileFormatInternal();
		isUnencryptedPDF = isUnencryptedPDFInternal();
		wasOpenedFromBuffer = false;
	}

	public MuPDFCore(Context context, byte buffer[], String magic) throws Exception {
		fileBuffer = buffer;
		globals = openBuffer(magic != null ? magic : "");
		if (globals == 0)
		{
			throw new Exception("Não é possível abrir o buffer");
		}
		file_format = fileFormatInternal();
		isUnencryptedPDF = isUnencryptedPDFInternal();
		wasOpenedFromBuffer = true;
	}

	public int countPages()
	{
		if (numPages < 0)
			numPages = countPagesSynchronized();
		return numPages;
	}

	private synchronized int countPagesSynchronized() {
		return countPagesInternal();
	}

	/* Shim function */
	private void gotoPage(int page)
	{
		if (page > numPages-1)
			page = numPages-1;
		else if (page < 0)
			page = 0;
		gotoPageInternal(page);
		this.pageWidth = getPageWidth();
		this.pageHeight = getPageHeight();
	}

	public synchronized void onDestroy() {
		destroying();
		globals = 0;
	}

	public synchronized RectF [] searchPage(int page, String text) {
		gotoPage(page);
		return searchPage(text);
	}

	public synchronized byte[] html(int page) {
		gotoPage(page);
		return textAsHtml();
	}

	public synchronized TextWord [][] textLines(int page) {
		gotoPage(page);
		TextChar[][][][] chars = text();

		// The text of the page held in a hierarchy (blocks, lines, spans).
		// Currently we don't need to distinguish the blocks level or
		// the spans, and we need to collect the text into words.
		ArrayList<TextWord[]> lns = new ArrayList<TextWord[]>();

		for (TextChar[][][] bl: chars) {
			if (bl == null)
				continue;
			for (TextChar[][] ln: bl) {
				ArrayList<TextWord> wds = new ArrayList<TextWord>();
				TextWord wd = new TextWord();

				for (TextChar[] sp: ln) {
					for (TextChar tc: sp) {
						if (tc.c != ' ') {
							wd.Add(tc);
						} else if (wd.w.length() > 0) {
							wds.add(wd);
							wd = new TextWord();
						}
					}
				}

				if (wd.w.length() > 0)
					wds.add(wd);

				if (wds.size() > 0)
					lns.add(wds.toArray(new TextWord[wds.size()]));
			}
		}

		return lns.toArray(new TextWord[lns.size()][]);
	}

	public synchronized void save() {
		saveInternal();
	}

}

File: ebook/src/main/java/com/artifex/mu/MuPDFCore.kt
package com.artifex.mu

import android.content.Context
import android.graphics.Bitmap
import android.graphics.PointF
import android.graphics.RectF

class MuPDFCore {
    /* Readable members */
    private var numPages = -1
    private var pageWidth = 0f
    private var pageHeight = 0f
    private var globals: Long
    private lateinit var fileBuffer: ByteArray
    private var file_format: String
    private var isUnencryptedPDF: Boolean
    private val wasOpenedFromBuffer: Boolean
    private external fun openFile(filename: String): Long
    private external fun openBuffer(magic: String): Long
    private external fun fileFormatInternal(): String
    private val isUnencryptedPDFInternal: Boolean
        private external get

    private external fun countPagesInternal(): Int
    private external fun gotoPageInternal(localActionPageNum: Int)
    private external fun getPageWidth(): Float
    private external fun getPageHeight(): Float
    private external fun drawPage(
        bitmap: Bitmap,
        pageW: Int, pageH: Int,
        patchX: Int, patchY: Int,
        patchW: Int, patchH: Int,
        cookiePtr: Long
    )

    private external fun updatePageInternal(
        bitmap: Bitmap,
        page: Int,
        pageW: Int, pageH: Int,
        patchX: Int, patchY: Int,
        patchW: Int, patchH: Int,
        cookiePtr: Long
    )

    private external fun searchPage(text: String?): Array<RectF>
    private external fun text(): Array<Array<Array<Array<TextChar>>>>
    private external fun textAsHtml(): ByteArray
    private external fun addMarkupAnnotationInternal(quadPoints: Array<PointF>, type: Int)
    private external fun addInkAnnotationInternal(arcs: Array<Array<PointF>>)
    private external fun deleteAnnotationInternal(annot_index: Int)
    private external fun passClickEventInternal(page: Int, x: Float, y: Float): Int
    private external fun setFocusedWidgetChoiceSelectedInternal(selected: Array<String>)
    private val focusedWidgetChoiceSelected: Array<String?>?
        private external get
    private val focusedWidgetChoiceOptions: Array<String?>?
        private external get
    private val focusedWidgetSignatureState: Int
        private external get

    private external fun checkFocusedSignatureInternal(): String?
    private external fun signFocusedSignatureInternal(keyFile: String, password: String): Boolean
    private external fun setFocusedWidgetTextInternal(text: String): Int
    private external fun getWidgetAreasInternal(page: Int): Array<RectF?>?
    private external fun getAnnotationsInternal(page: Int): Array<Annotation?>?

    //private native OutlineItem [] getOutlineInternal();
    private external fun hasOutlineInternal(): Boolean
    private external fun needsPasswordInternal(): Boolean
    private external fun authenticatePasswordInternal(password: String): Boolean
    private external fun startAlertsInternal()
    private external fun stopAlertsInternal()
    private external fun destroying()
    private external fun hasChangesInternal(): Boolean
    private external fun saveInternal()
    private external fun createCookie(): Long
    private external fun destroyCookie(cookie: Long)
    private external fun abortCookie(cookie: Long)
    private external fun startProofInternal(resolution: Int): String?
    private external fun endProofInternal(filename: String)
    private external fun getNumSepsOnPageInternal(page: Int): Int
    private external fun controlSepOnPageInternal(page: Int, sep: Int, disable: Boolean): Int
    private external fun getSepInternal(page: Int, sep: Int): Separation?
    external fun javascriptSupported(): Boolean
    inner class Cookie {
        private val cookiePtr: Long

        init {
            cookiePtr = createCookie()
            if (cookiePtr == 0L) throw OutOfMemoryError()
        }

        fun abort() {
            abortCookie(cookiePtr)
        }

        fun destroy() {
            // We could do this in finalize, but there's no guarantee that
            // a finalize will occur before the muPDF context occurs.
            destroyCookie(cookiePtr)
        }
    }

    constructor(context: Context?, filename: String) {
        globals = openFile(filename)
        if (globals == 0L) {
            throw Exception(String.format("Não é possível abrir o item selecionado", filename))
        }
        file_format = fileFormatInternal()
        isUnencryptedPDF = isUnencryptedPDFInternal
        wasOpenedFromBuffer = false
    }

    constructor(context: Context?, buffer: ByteArray, magic: String?) {
        fileBuffer = buffer
        globals = openBuffer(magic ?: "")
        if (globals == 0L) {
            throw Exception("Não é possível abrir o buffer")
        }
        file_format = fileFormatInternal()
        isUnencryptedPDF = isUnencryptedPDFInternal
        wasOpenedFromBuffer = true
    }

    fun countPages(): Int {
        if (numPages < 0) numPages = countPagesSynchronized()
        return numPages
    }

    @Synchronized
    private fun countPagesSynchronized(): Int {
        return countPagesInternal()
    }

    /* Shim function */
    private fun gotoPage(page: Int) {
        var page = page
        if (page > numPages - 1) page = numPages - 1 else if (page < 0) page = 0
        gotoPageInternal(page)
        pageWidth = getPageWidth()
        pageHeight = getPageHeight()
    }

    @Synchronized
    fun onDestroy() {
        destroying()
        globals = 0
    }

    @Synchronized
    fun searchPage(page: Int, text: String?): Array<RectF> {
        gotoPage(page)
        return searchPage(text)
    }

    @Synchronized
    fun html(page: Int): ByteArray {
        gotoPage(page)
        return textAsHtml()
    }

    @Synchronized
    fun textLines(page: Int): Array<Array<TextWord>> {
        gotoPage(page)
        val chars = text()

        // The text of the page held in a hierarchy (blocks, lines, spans).
        // Currently we don't need to distinguish the blocks level or
        // the spans, and we need to collect the text into words.
        val lns = ArrayList<Array<TextWord>>()
        for (bl in chars) {
            if (bl == null) continue
            for (ln in bl) {
                val wds = ArrayList<TextWord>()
                var wd = TextWord()
                for (sp in ln) {
                    for (tc in sp) {
                        if (tc.c != ' ') {
                            wd.Add(tc)
                        } else if (wd.w.length > 0) {
                            wds.add(wd)
                            wd = TextWord()
                        }
                    }
                }
                if (wd.w.length > 0) wds.add(wd)
                if (wds.size > 0) lns.add(wds.toTypedArray())
            }
        }
        return lns.toTypedArray()
    }

    @Synchronized
    fun save() {
        saveInternal()
    }

    companion object {
        /* load our native library */
        private var gs_so_available = false

        init {
            println("Loading dll")
            System.loadLibrary("mupdf_java")
            println("Loaded dll")
            if (gprfSupportedInternal()) {
                try {
                    System.loadLibrary("gs")
                    gs_so_available = true
                } catch (e: UnsatisfiedLinkError) {
                    gs_so_available = false
                }
            }
        }

        /* The native functions */
        private external fun gprfSupportedInternal(): Boolean
    }
}